tests/migration/postcopy.go

```diff
@@ -0,0 +1,412 @@
+						Eventually(func() v1.VirtualMachineInstanceMigrationPhase {
+							migration, err := virtClient.VirtualMachineInstanceMigration(migration.Namespace).Get(migration.Name, &metav1.GetOptions{})
+							Expect(err).ToNot(HaveOccurred())
+
+							phase := migration.Status.Phase
+							Expect(phase).NotTo(Equal(v1.MigrationSucceeded))
+
+							return phase
+
+						}, 150, 1*time.Second).Should(Equal(v1.MigrationFailed))
+					}
+
+					It("should make sure that VM restarts after failur", func() {
+
+						By("creating a large VM with RunStrategyRerunOnFailure")
+						vm := createLargeVirtualMachine(testsuite.NamespacePrivileged)
+
+						// update the migration policy to ensure slow pre-copy migration progress instead of an immidiate cancelation.
+						migrationPolicy.Spec.CompletionTimeoutPerGiB = kvpointer.P(int64(20))
+						migrationPolicy.Spec.BandwidthPerMigration = kvpointer.P(resource.MustParse("1Mi"))
+						applyKubevirtCR()
+
+						By("Starting the VirtualMachine")
+						vm = tests.RunVMAndExpectLaunchWithRunStrategy(virtClient, vm, v1.RunStrategyRerunOnFailure)
+						vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &metav1.GetOptions{})
+						Expect(err).ToNot(HaveOccurred())
+
+						By("Checking that the VirtualMachineInstance console has expected output")
+						Expect(console.LoginToFedora(vmi)).To(Succeed())
+
+						// Need to wait for cloud init to finish and start the agent inside the vmi.
+						Eventually(matcher.ThisVMI(vmi), 12*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
+
+						runStressTest(vmi, "350M", stressDefaultSleepDuration)
+
+						By("Starting the Migration")
+						migration := libmigration.New(vmi.Name, vmi.Namespace)
+						migration = libmigration.RunMigration(virtClient, migration)
+
+						// check VMI, confirm migration state
+						waitUntilMigrationMode(vmi, v1.MigrationPostCopy, 300)
+
+						// launch killer pod on every node that isn't the vmi's node
+						By("Starting migration killer pods")
+						runMigrationKillerPod(vmi.Status.NodeName)
+
+						By("Making sure that post-copy migration failed")
+						observePostCopyMigrationFailure(migration)
+
+						By("Removing migration killer pods")
+						removeMigrationKillerPod()
+
+						By("Ensuring the VirtualMachineInstance is restarted")
+						Eventually(ThisVMI(vmi), 240*time.Second, 1*time.Second).Should(beRestarted(vmi.UID))
+					})
+				})
+			})
+		})
+	})
+})
+
+func beRestarted(oldUID types.UID) gomegatypes.GomegaMatcher {
+	return gstruct.PointTo(gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+		"ObjectMeta": gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+			"UID": Not(Equal(oldUID)),
+		}),
+		"Status": gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+			"Phase": Equal(v1.Running),
+		}),
+	}))
+}
```

Comment:
This function is great and it's the second time I see it. Maybe we should start thinking of creating a new file under `matcher` pkg.